BDMS 233: transfer updates

jacob-a-brown opened the pull request:
<!--
- Title the PR with the JIRA project & ticket number and short description (e.g., BDMS-123: Example display bug), or NO TICKET
- Keep sections short and scannable.
-->
### **Why**
_This PR addresses the following problem / context:_
- notes need to be transferred to the polymorphic notes table

### **How**
_Implementation summary - the following was changed / added / removed:_
- updated well and location transfer scripts/code to transfer the following note types:
  - well: `Other`, `Water`, `Construction`, `Measuring`
  - location: `Other`, `Coordinate`
- refactored legacy notes fields to go into the into the polymorphic notes table:
  - `thing.well_construction_notes` --> `note_type = "Construction"`
  - `location.nma_notes_location` -- > `note_type = "Other"`
  - `location.nma_coordinate_notes` --> `note_type = "Coordinate"`
- update `well-notes.py` feature test for the refactoring of `well_construction_notes` to the polymorphic notes table
  

### **Notes**
_Any special considerations, workarounds, or follow-up work to note?_
- `Measuring` notes for a `thing` come from `WaterLevels.SiteNotes` in **NM_Aquifer**. The documentation for that field states "Optional notes concerning the nature of the well site or measuring event."
- This PR utilizes work done for **well-additional-information** so it should not be merged until PR #257  and PR #258 are merged into staging. I am opening a PR, though, to preserve the work done and get feedback when able. That being said, once those two PRs are merged into staging the number of files (and lines) changed will decrease dramatically.
  - I'll re-request reviews once those two PRs are merged.

jacob-a-brown committed:
WIP: first stab at well additional information

jacob-a-brown committed:
Merge branch 'staging' into bdms-227-jab-bdms-229

jacob-a-brown committed:
WIP: well additional information bdd tests

jacob-a-brown committed:
feat: use function to retrieve polymoprhic records

jacob-a-brown committed:
feat: pass bdd test for well completion date

jacob-a-brown committed:
feat: pass well driller name bdd test

jacob-a-brown committed:
feat: implement well construction method

jacob-a-brown committed:
feat: implement well casing diameter in inches

jacob-a-brown committed:
feat: implement well pump type

jacob-a-brown committed:
feat: implement well pump depth

jacob-a-brown committed:
Merge branch 'bdms-227' into bdms-227-jab-bdms-229

jacob-a-brown committed:
Merge branch 'bdms-227' into bdms-227-jab-updates-to-pass-tests

jacob-a-brown committed:
fix: remove duplicate lexicon

jacob-a-brown committed:
refactor: test for casing materials

jacob-a-brown committed:
Merge branch 'bdms-227-jab-bdms-229' into bdms-227-jab-updates-to-pass-tests

jacob-a-brown committed:
feat: update tests to include well casing materials

jacob-a-brown committed:
fix: fix casing materials test

ksmuczynski committed:
Merge branch 'kas-bdms-221-225-core-well-info-models-schemas' into kas-227-231-additional-well-info-models

# Conflicts:
#	.pre-commit-config.yaml
#	core/lexicon.json
#	tests/features/environment.py

jacob-a-brown committed:
fix: uncomment pytest from pre commit

jacob-a-brown committed:
refactor: rename Permission -> PermissionHistory | move mixin to same file

jacob-a-brown committed:
feat: add permission types to lexicon

jacob-a-brown committed:
refactor: import PermissionHistoryMixin from correct dir

jacob-a-brown committed:
fix: fix imports for newly renamed PermissionHistory

jacob-a-brown committed:
feat: add permission_type to lexicon

jacob-a-brown committed:
feat: update util to correspond with bdms 221

jacob-a-brown committed:
fix: sync bdms 221/227 for util functions

jacob-a-brown committed:
feat: implement permissions

jacob-a-brown committed:
feat: add is_suitable_for_datalogger to thing

ksmuczynski committed:
feat: add 'AquiferSystem' model with relationships and controlled vocabulary fields

- Implemented `AquiferSystem` model for master reference of aquifer systems and hydrogeologic units
- Added table index for aquifer system name
- Included categories and terms for aquifer type and significance level

ksmuczynski committed:
refactor: remove completed TODO about lexicon updates

ksmuczynski committed:
refactor: clarify relationship  name and add association proxy

Added an association proxy, `things`, to the `AquiferSystem` model. This provides direct access to the `Thing` objects.

ksmuczynski committed:
refactor: clarify proxy purpose doc statement

ksmuczynski committed:
feat: add ThingAquiferAssociation model and update Thing relationships

- Created ThingAquiferAssociation model to manage many-to-many relationships between Thing and AquiferSystem
- Updated Thing model to include aquifer_associations relationship for linking to aquifer systems

ksmuczynski committed:
feat: add `aqufiers` proxy to Thing model

Added an `aquifers` proxy to the Thing model to directly access the AqufierSystems associated with a particular Thing.

ksmuczynski committed:
feat: Link WellScreen to AquiferSystem

This commit addresses the need to handle scenarios where multiple well screens may target different aquifers.

Added the aquifer_system_id foreign key to the WellScreen
model.

This creates a direct, physical link between a screened interval and
the specific hydrogeologic unit it monitors.

ksmuczynski committed:
feat: add GeologicFormation and ThingFormationAssociation models with Thing relationship updates

- Introduced GeologicFormation model as a master reference for geologic formations
- Created ThingFormationAssociation model to manage many-to-many relationships between Thing and GeologicFormation, including depth interval fields
- Updated Thing model to include formation_associations relationship and association proxy for direct access to related geologic formations

ksmuczynski committed:
feat(lexicon): add "formation_code" category and associated terms to lexicon

ksmuczynski committed:
rafactor(model): Remove TODO about adding formation_code values to lexicon

ksmuczynski committed:
refactor(lexicon): refine terms associated with "geographic_scale".

ksmuczynski committed:
refactor(model): Replace hardcoded srid with SRID_WGS84 from constants

Replace the hardcoded `srid` in the `aquifer_system` and `geologic_formation` models with the SRID_WGS84 from constants.

ksmuczynski committed:
refactor(model): Add geologic_formation foreign key to WellScreen model.

ksmuczynski committed:
refactor(model): enhance relationships for `ThingAquiferAssociation` and `ThingFormationAssociation` models

Add `lazy="joined"` to `ThingAquiferAssociation` and `ThingFormationAssociation` models to eagerly load relationships.

ksmuczynski committed:
refactor(model): refactor `lithology` field to a lexicon_term

- Refactor `lithology` field to a lexicon_term
- Add lithology values to lexicon

ksmuczynski committed:
feat(core): create placeholder `formations.json`

This is a placeholder file to allow the API to map formation names to codes. It can be  periodically updated from the authoritative source (.e.g USGS).

ksmuczynski committed:
refactor: add aquifer and geology related models to `db/__init__.py`

ksmuczynski committed:
feat: add aquifer and geology related enums.

ksmuczynski committed:
feat(schemas): add response schemas for aquifer systems and geologic formations

Add comprehensive response schemas for AquiferSystem and GeologicFormation
models. Integrate these schemas into
Thing/Well response models to provide complete geological context.

Changes:
-  Add AquiferSystemResponse with name, type, scale, and boundary fields
- Add GeologicFormationResponse with formation code, lithology, and boundary
- Add ThingFormationAssociationResponse with depth interval data
- Update WellResponse to include aquifers and formations lists
- Add field validators to convert ORM association objects to response schemas
- Update WellScreenResponse to include aquifer_system and geologic_formation
- Update CreateWellScreen and UpdateWellScreen with geological associations

jacob-a-brown committed:
Merge pull request #237 from DataIntegrationGroup/bdms-227-jab-bdms-229

BDMS 227: well additional information tests

jacob-a-brown committed:
Merge branch 'staging' into bdms-227

jacob-a-brown committed:
Merge branch 'bdms-227' into bdms-227-jab-updates-to-pass-tests

jacob-a-brown committed:
fix: fix artifacts from merge conflicts

ksmuczynski committed:
refactor: rename `ThingFormationAssociation` usages to `ThingGeologicFormationAssociation`

ksmuczynski committed:
refactor: rename `ThingFormationAssociation` file`ThingGeologicFormationAssociation`

ksmuczynski committed:
refactor(schema): remove aquifer and formation field validators from `schemas/thing.py`

jacob-a-brown committed:
feat: implement data source tests for well additional information

jacob-a-brown committed:
fix: remove outdated notes

jacob-a-brown committed:
feat: transfer water level sample permissions

jacob-a-brown committed:
feat: transfer permissions from legacy db

jacob-a-brown committed:
feat: transfer well construction information

jacob-a-brown committed:
WIP: transfers for well additional info

jacob-a-brown committed:
feat: implement origin_type

jacob-a-brown committed:
Merge branch 'bdms-227-jab-updates-to-pass-tests' into bdms-227-transfer-updates

jacob-a-brown committed:
feat: transfer well source information

jacob-a-brown committed:
feat: add Quemado Mutual Water and Sewage Works Association to organizations for contacts

jacob-a-brown committed:
fix: remove erroneous or check

jacob-a-brown committed:
feat: transfer well pump type

parse the construction notes field to get pump types. this can be more
sophisticated, but that's a future problem. for now just check for some
basic pump type string patterns

jacob-a-brown committed:
fix: change Air-rotary to Air-Rotary to correspond with lu table

This can be updated later on but should be "Air-Rotary" for the transfer

ksmuczynski committed:
feat(model): Add `AquiferType` model and rename `aquifer_type` to `primary_aquifer_type`

- Add AquiferType model to store multiple aquifer characteristics per association
- Rename AquiferSystem.aquifer_type to primary_aquifer_type for clarity
- Update ThingAquiferAssociation with aquifer_types relationship
- Update related schemas to use primary_type

This enables capturing wells that encounter multiple aquifer characteristics
(e.g., both fractured and confined) without compound naming.

ksmuczynski committed:
Merge branch 'bdms-227' into kas-227-231-additional-well-info-models

# Conflicts:
#	.pre-commit-config.yaml
#	core/enums.py
#	core/lexicon.json
#	db/__init__.py
#	schemas/thing.py
#	tests/features/environment.py
#	tests/features/steps/well-additional-information.py

ksmuczynski committed:
feat(model): add eager loading

Add eager loading to aquifer and geology relationships to the `thing` model.

ksmuczynski committed:
feat(schema): refactor GeoJSON responses for `aquifer_system` and `geologic_formation`

jacob-a-brown committed:
Merge pull request #248 from DataIntegrationGroup/kas-227-231-additional-well-info-models

BDMS-227-231-aquifer-geology-models

jacob-a-brown committed:
Merge branch 'bdms-227' into bdms-227-jab-updates-to-pass-tests

jacob-a-brown committed:
fix: fix __table_args__ for aquifer and geology

jacob-a-brown committed:
fix: add geologic_formation relation to WellScreen

This was missing on the WellScreen side of the relationship

jacob-a-brown committed:
fix: add missing comma

jacob-a-brown committed:
fix: use BaseResponseModel for GeoJSONProperties in location schema

jacob-a-brown committed:
fix: include baseresponseinfo in location geojson properties

jacob-a-brown committed:
feat: implement aquifer_systems in thing response

jacob-a-brown committed:
feat: fix aquifer systems test

jacob-a-brown committed:
fix: return aquifer system and types in dict

this allows us to couple types to a system incase a well is in
multiple aquifers

jacob-a-brown committed:
feat: implement geologic_formations from feature files

jacob-a-brown committed:
Merge branch 'bdms-227-jab-updates-to-pass-tests' into bdms-227-transfer-updates

jacob-a-brown committed:
fix: remove duplicate well pump types in lexicon

this was an artifact from merge conflicts

jacob-a-brown committed:
fix: use OpenWellLoggerOK for is_suitable_for_datalogger field

Laila indicated that this just describes if a datalogger can be
physically installed at the well. It does not pertain to
permissions

jacob-a-brown committed:
feat: update CreateWell for transfers

jacob-a-brown committed:
fix: return aquifer/geology str instead of full response

the well screen response includes the aquifer and geology information,
but just as strings of names and codes instead of the full nested response objects.

jacob-a-brown committed:
refactor: clean up permission history testing data

use a for loop for cleaner and more maintainable code

jacob-a-brown committed:
fix: use origin_type to retrieve well depth source

jacob-a-brown committed:
refactor: use __tabename__ for polymorphic mixins

this ensures fidelity in table naming across all database models

jacob-a-brown committed:
fix: use __tablename__ for NotesMixin add_note

Now that __tablename__ is used throughout it should be
used for the target_table in add_note

jacob-a-brown committed:
Merge branch 'bdms-227-jab-updates-to-pass-tests' into bdms-227-transfer-updates

jacob-a-brown committed:
refactor: move well_construction_notes to polymorphic table

all notes should reside in that table rather than as separate columns in the well table

jacob-a-brown committed:
feat: transfer other, water, and construction notes for wells

jacob-a-brown committed:
feat: transfer Measuring notes

transfer from WaterLevels.SiteNotes to polymorphic notes table
where target_table = 'thing' and the target_id is thing.id

jacob-a-brown committed:
feat: transfer location notes

jirhiker commented on tests/features/steps/well-additional-information.py:
```diff
@@ -0,0 +1,210 @@
+        assert "aquifer_system" in aquifer
+    assert sorted(
+        [a.get("aquifer_system") for a in context.water_well_data["aquifers"]]
+    ) == sorted([system.name for system in context.objects["aquifer_systems"]])
```
comparing sets might be a little more elegant

`{a.get("aquifer_system") for a in ... } == {system.name for system in ... }`

jacob-a-brown replied:
This will get changed in PR #256 and then will be present in this PR once that one is merged.

jirhiker commented on transfers/permissions_transfer.py:
```diff
@@ -0,0 +1,87 @@
+        allow_water_level_samples = wdf.loc[
+            wdf["PointID"] == well.name, "MonitorOK"
+        ].values
+        if len(allow_water_level_samples) > 0 and allow_water_level_samples is not None:
```
you want to test if `allow_water_level_samples` is not None before testing its length.  this will fail if `allow_water_level_samples` is None because you cannot `len(None)`

jacob-a-brown replied:
This will get changed in PR #257 and then will be present in this PR once that one is merged.

jacob-a-brown replied:
See fix in PR #263

jirhiker commented on transfers/permissions_transfer.py:
```diff
@@ -0,0 +1,87 @@
+    wdf = read_csv("WellData", dtype={"OSEWelltagID": str})
+    wdf = replace_nans(wdf)
+
+    transferred_wells = (
```
all the other transfers use `filter_to_valid_point_ids` for this functionality

jacob-a-brown replied:
The permissions require the `id` of the `thing` record to set the `target_id` field, which is why the wells are queried in `permissions_transfer.py` (after the wells have been transferred and assigned `id`s by the database). Using `filter_to_valid_point_ids` would get me the correct `thing.name` for all of the wells, but wouldn't fetch the `id`s.

jacob-a-brown replied:
When I was figuring out this transfer it made me sense to me to do it this way, but I can move this to `well_transfer.py` after line 365 - the second after the comment "add things that need  well id"

jacob-a-brown replied:
Upon reflection I think it makes more sense to go there. It'll be better organized and easier to follow

jacob-a-brown replied:
I forgot another reason why `permission_transfer.py` is its own file... The permissions require an associated `contact`, so they need to be transferred after both the `things` and `contacts` are transferred (which is why they can't go into `well_transfer.py` (which negates the above two comments)

jirhiker commented on transfers/thing_transfer.py:
```diff
@@ -57,12 +57,17 @@ def transfer_thing(session: Session, site_type: str, make_payload, limit=None) -
             )
+            for note_type, note_content in location_notes.items():
+                if not isna(note_content):
+                    location.add_note(note_content, note_type)
```
Doesn't add_note just create a note object? you still need to add it to the session

jacob-a-brown replied:
I just looked at it and it does - I'll fix to add it to the session

jirhiker commented on transfers/waterlevels_transfer.py:
```diff
@@ -385,6 +385,19 @@ def transfer_water_levels(session):
+
+            # WaterLevels.SiteNotes --> notes where note_type = "measuring_notes"
+            if not pd.isna(row.SiteNotes):
+                note = thing.add_note(
```
how is this note going to be associated with this water level measurement?

jacob-a-brown replied:
It won't be associated with water level measurements in Ocotillo, but with wells. The step from `well-notes.feature` indicates `the response should include measuring notes (notes about measuring/visiting the well, on Access form)`. This is really the only field in the database that explicitly records notes about measures/visits, so it is now being associated with a well and not an observation.

The documentation is: "Optional notes concerning the nature of the well site or measuring event."

chasetmartin replied:
I have messages out to Ethan and Sianin to clarify "measuring notes" since this field they mention that they want to see and enter data into is not found specifically in NM Aquifer and needs clarification. Looking at some of the values in WaterLevels.SiteNotes, they are definitely observation level, containing info like battery levels, which seems most relevant when associated with a specific time/date of observation.
<img width="1594" height="419" alt="image" src="https://github.com/user-attachments/assets/d4f145bc-5f21-4470-ac0e-fb6bd0630dc2" />
I assume this response field from the old AMP API is the WaterLevels.SiteNotes?

jacob-a-brown replied:
That's correct - that response comes from `WaterLevels.SiteNotes`

jirhiker replied:
as @chasetmartin mentioned WaterLevel.SiteNotes is related to observation level notes where time/date info is necessary. 

I think the feature is saying return all the observation notes for this well, i.e. all the WaterLevel.SiteNotes.  So WaterLevel.SiteNotes needs to be associated with a measurement. The API then collects all these notes and returns them with the other well metadata

jacob-a-brown replied:
Got it. I'll then update this transfer, as well as update the feature file/test to reflect this

jirhiker commented on transfers/well_transfer.py:
```diff
@@ -117,6 +118,21 @@ def _extract_casing_materials(row) -> list[str]:
 
+def _extract_well_pump_type(row) -> str | None:
+    construction_notes = row.ConstructionNotes.lower()
+    if "pump" in construction_notes:
```
a regex would be better here. for the most part this will work but what about the note.   "Handle on steel tape broke. Needed to jettison our tools. Pump is a submersible".  The function would incorrectly identify the pump as "Jet". 

```
import re


pattern = re.compile(
    r"\b(?P<term>jet|hand|submersible)\b|\b(?P<phrase>line[-\s]+shaft)\b",
    re.IGNORECASE
)

def first_matched_term(text: str):
    m = pattern.search(text)
    if not m:
        return None
    return m.group("term") or m.group("phrase")

PUMP_MAPPING = {
    "jet": "Jet",
    "hand": "Hand",
    "submersible": "Submersible"
}

def _extract_well_pump_type(row) -> str | None:
    construction_notes = row.ConstructionNotes.lower()
    return PUMP_MAPPING.get(first_matched_term(construction_notes), None)
```

jacob-a-brown replied:
See fix in PR #263

jirhiker commented on transfers/well_transfer.py:
```diff
@@ -236,6 +255,9 @@ def transfer_wells(session: Session, flags: dict = None, limit: int = 0) -> None
                 [] if isna(row.CasingDescription) else _extract_casing_materials(row)
             )
+            well_pump_type = (
+                _extract_well_pump_type(row) if row.ConstructionNotes else None
```
move the `if row.ConstructionNotes` test to `_extract_well_pump_type`

jacob-a-brown replied:
This will get changed in PR #257 and then will be present in this PR once that one is merged.

jirhiker commented on .pre-commit-config.yaml:
```diff
@@ -16,15 +16,15 @@ repos:
-        types: [python] # Specify relevant file types for your tests
-        pass_filenames: false
-        always_run: true
+#  - repo: local
```
You should not be committing these temporary, dev changes to .pre-commit-config.yaml

jirhiker reviewed (commented):
Looks good. I will need to refactor slightly to match the latest transfer branch

jacob-a-brown committed:
refactor: use sets in feature tests for comparison

jacob-a-brown committed:
Merge branch 'bdms-227-jab-updates-to-pass-tests' into bdms-227-transfer-updates

jacob-a-brown committed:
Merge branch 'bdms-227-transfer-updates' into bdms-233-transfer-updates

jacob-a-brown committed:
refactor: return permission history records for a well

instead of a specific permission type, return a list of permission records for all permission types associated with the well.

jacob-a-brown committed:
Merge branch 'bdms-227-jab-updates-to-pass-tests' into bdms-227-transfer-updates

jacob-a-brown committed:
Merge branch 'bdms-227-transfer-updates' into bdms-233-transfer-updates

jacob-a-brown committed:
Merge pull request #256 from DataIntegrationGroup/bdms-227-jab-updates-to-pass-tests

well additional information testing/implementation updates

jacob-a-brown committed:
Merge branch 'bdms-227' into bdms-227-transfer-updates

jacob-a-brown committed:
Merge branch 'bdms-227-transfer-updates' into bdms-233-transfer-updates

jacob-a-brown committed:
Merge pull request #257 from DataIntegrationGroup/bdms-227-transfer-updates

Bdms 227 transfer updates

jacob-a-brown committed:
Merge branch 'bdms-227' into bdms-233-transfer-updates

jacob-a-brown committed:
fix: use regex to check for pump type in well construction notes

regex is a more robust way to check for the presence of pump types in construction notes, handling variations in spacing and punctuation.

jacob-a-brown committed:
fix: ensure permissions are not null during transfer

check to make sure that permissions are set before transferring them
from one entity to another to avoid potential errors.

jacob-a-brown committed:
refactor: check for existence of permissions before converting

jacob-a-brown committed:
Merge branch 'staging' into bdms-227

jacob-a-brown committed:
Merge branch 'bdms-227' into bdms-227-transfer-updates

jacob-a-brown committed:
Merge branch 'bdms-227-transfer-updates' into bdms-233-transfer-updates

jacob-a-brown committed:
refactor: use NotesMixin for FieldEvent

use the polymorphic table rather than a standalone field

jacob-a-brown committed:
fix: import NotesMixin correctly for FieldEvent

jacob-a-brown committed:
refactor: rename Other to General for notes

This is the name used for serialization so there should be
correspondence for ease of use and maintenance

jacob-a-brown committed:
feat: add SiteNotes for field events
